### PR TITLE
Features/basic infrastructure

### PR DESCRIPTION
**Pull request by VasylKosiv**

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+BIND_IP=0.0.0.0
+BIND_PORT=8000
+BIND_DEBUG_PORT=8001
+CONSUL_ADDRESS=consul:8500
+
+#Gralog environment variable
+GRAYLOG_PASSWORD_SECRET=usermanagerlvl48
+# Password: gopher
+GRAYLOG_ROOT_PASSWORD_SHA2=9cc1ee455a3363ffc504f40006f70d0c8276648a5d3eb3f9524e94d1b7a83aef
+GRAYLOG_HTTP_EXTERNAL_URI=http://127.0.0.1:9000/
```
Comment (crnopster) [inline .env]: It is not a good idea to push files like this.
Add .env to .gitignore

File: config/consul_registr/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,5 @@
+FROM consul
+
+COPY registr.json /consul/configs/
+
+ENTRYPOINT ["consul", "agent", "-config-dir=/consul/configs", "-data-dir=/tmp/consul"]
```
Comment (crnopster) [inline config/consul_registr/Dockerfile]: Move this to infra/consul
It is not a part of positioning-service config

File: config/consul_registr/registr.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        "port": 9000,
+        "checks": [
+            {
+                "name": "graylog-check-9000",
+                "tcp": "graylog:9000",
+                "interval": "30s",
+                "timeout": "5s"
+            }]
+        }]
+    }
```
Comment (crnopster) [inline config/consul_registr/registr.json]: Move this to infra/consul
It is not a part of positioning-service config

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    depends_on: 
+      - db
+
+
+
+  consul-agent-1: &consul-agent
+    image: consul:1.7.1
+    networks:
+      - gpsnet
+    command: "agent -retry-join consul-server-bootstrap -client 0.0.0.0"
```
Comment (crnopster) [inline docker-compose.yml]: What does this agent do?

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+COPY server          server
+COPY storage         storage
+COPY go.*            ./
+
+WORKDIR /opt/resource/cmd/umserver
+RUN go build -o /opt/services/example-app .
+
+FROM alpine:3.7
+COPY --from=builder /opt/services/example-app /opt/services/example-app
+CMD /opt/services/example-app
```
Comment (crnopster) [inline Dockerfile]: example-app is not good name for your service

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,16 @@
+up: ##build and run project in docker conteiner
+	@docker-compose up --build -d
+
+
+
+
+test: ##run test
+	@go test -v ./...
```
Comment (crnopster) [inline Makefile]: tests can be run in Dockerfile

File: .env (hunk 1 above)
Comment (VasylKosiv) [reply .env]: Fix problem

File: Dockerfile (hunk 5 above)
Comment (VasylKosiv) [reply Dockerfile]: rename service

File: config/consul_registr/Dockerfile (hunk 2 above)
Comment (VasylKosiv) [reply config/consul_registr/Dockerfile]: Done

File: config/consul_registr/registr.json (hunk 3 above)
Comment (VasylKosiv) [reply config/consul_registr/registr.json]: Done

File: docker-compose.yml (hunk 4 above)
Comment (VasylKosiv) [reply docker-compose.yml]: Remove this agent

File: Makefile (hunk 6 above)
Comment (VasylKosiv) [reply Makefile]: Agree del this comand

File: README.md [hunk 7]
```diff
@@ -24,6 +24,13 @@ User is the central model of the service:
 
 Service will use PostgreSQL as a storage for user data. All passwords have to be saved securely using hashing algorithms and saults. Access rights for the service SQL user have to be exactly the same that required to cover needed queries requirements. Database connection have to be able to recover after disconnect.
 
+###### Database migration
+PostgreSQL database is migrate from oficial image postgres:9.6.17. Initial structure of database is done with "Database migrations. CLI and Golang library." https://github.com/golang-migrate/migrate using Docker image for golang-migrate binary  https://hub.docker.com/r/migrate/migrate/.
+Queries are placed in folder "migrations" and must named as N____.up.sql or N___.down.sql, where N means the order of executing.
+Execution of migration
+    #docker-compose up -d db
+    #docker-compose up -d
```
Comment (crnopster) [inline README.md]: So we could run all services in two commands instead of using one?

Comment (crnopster) [inline README.md]: Check markdown syntax. This part is incorrect

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+COPY --from=modules /go/pkg/ /go/pkg/
+
+WORKDIR /opt/resource/
+COPY cmd             cmd
+COPY config          config
+COPY middleware      middleware
+COPY model           model
+COPY server          server
+COPY storage         storage
+COPY go.*            ./
```
Comment (radikh) [inline Dockerfile]: Is it reasonable to add each package to the container instead of entire project?

File: README.md (hunk 7 above)
Comment (VasylKosiv) [reply README.md]: fix

Comment (VasylKosiv) [reply README.md]: fix

File: Dockerfile (hunk 8 above)
Comment (VasylKosiv) [reply Dockerfile]: Agree with you. Changed it

**Comment by radikh** [comment]
Just a reminder: comments should be resolved by those who left them.

File: .env.example [hunk 9]
```diff
@@ -0,0 +1,20 @@
+#Postgres environment variable
+
+POSTGRES_USER=postgres
+POSTGRES_PASSWORD=
+POSTGRES_DB=um_db
+UM_PASSWORD=
+
+
+#User-Manager environment variable
```
Comment (dmitriyminer) [inline .env.example]: These comments should be located in readme

File: infra/consul/registr.json [hunk 10]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (dmitriyminer) [inline infra/consul/registr.json]: Could you please format json

File: .env.example [hunk 11]
```diff
@@ -0,0 +1,20 @@
+#Postgres environment variable
+
+POSTGRES_USER=postgres
+POSTGRES_PASSWORD=
+POSTGRES_DB=um_db
+UM_PASSWORD=
+
```
Comment (dmitriyminer) [inline .env.example]: Please remove all extra lines

File: Makefile [hunk 12]
```diff
@@ -0,0 +1,17 @@
+docker_up: ##build and run project in docker conteiner
+	@docker-compose up --build -d
+
+test: ##run test
+	  ##This task for future		
+
```
Comment (dmitriyminer) [inline Makefile]: Extra lines

File: go.mod [hunk 13]
```diff
@@ -1,3 +1,5 @@
 module github.com/lvl484/user-manager
 
 go 1.13
+
+require github.com/gorilla/mux v1.6.2
```
Comment (dmitriyminer) [inline go.mod]: Do you have dependency to `github.com/gorilla/mux`?

Comment (VasylKosiv) [reply go.mod]: We don't need this dependency
but docker container didn't start without any dependency in  go.mod

File: .env.example (hunk 9 above)
Comment (VasylKosiv) [reply .env.example]: Add comment to README.md

File: infra/consul/registr.json (hunk 10 above)
Comment (VasylKosiv) [reply infra/consul/registr.json]: Done

File: .env.example (hunk 11 above)
Comment (VasylKosiv) [reply .env.example]: Remove all extra empty line

File: Makefile (hunk 12 above)
Comment (VasylKosiv) [reply Makefile]: Remove all extra empty line

File: go.mod (hunk 13 above)
Comment (VasylKosiv) [reply go.mod]: Insert comment in go.mod

Comment (radikh) [reply go.mod]: @VasylKosiv, what was the reason of failure?

File: Dockerfile (hunk 8 above)
Comment (radikh) [reply Dockerfile]: And now we will include everything to docker container?

File: go.mod (hunk 13 above)
Comment (VasylKosiv) [reply go.mod]: Previously, the container fell down on this command: COPY --from=modules /go/pkg/ /go/pkg/ from Docker file. When this depdendency was added, container started running without error.
But now, I'm trying to ran the docker-compose without this dependecy, to see type of error, and container is up without error. So I don't have reason to leve this dependency

File: Dockerfile (hunk 8 above)
Comment (VasylKosiv) [reply Dockerfile]: Add .dockerignore whith list of exceptions

File: go.mod (hunk 13 above)
Comment (radikh) [reply go.mod]: So add unused dependency is not a solution. If my understanding is right the problem was in non existing folder with dependencies when no dependencies are specified. So you'd better to solve that issue.

Comment (VasylKosiv) [reply go.mod]: I agree whith you. But is it so important in our case? Because our project will have some dependencies for sure.

Comment (VasylKosiv) [reply go.mod]: Maybe even this dependency

Comment (radikh) [reply go.mod]: It's important to be precise, at least it's enough to run `mkdir` to make a directory.

Comment (VasylKosiv) [reply go.mod]: Thank you for advice. Fix it.

File: Dockerfile [hunk 14]
```diff
@@ -0,0 +1,18 @@
+FROM golang:1.13 as modules
+ADD go.mod /m/
+RUN cd /m && go mod download && mkdir -p /go/pkg/
```
Comment (crnopster) [inline Dockerfile]: What will happen if "go mod download" will create folder before "mkdir" ?

Comment (VasylKosiv) [reply Dockerfile]: https://linux.die.net/man/1/mkdir

File: docker-compose.yml [hunk 15]
```diff
@@ -1,7 +1,57 @@
...
+    networks:
+      - gpsnet
+  
+  elasticsearch:
+    image:  docker.elastic.co/elasticsearch/elasticsearch:7.6.1
+    environment:
+      - http.host=0.0.0.0
+      - transport.host=localhost
+      - network.host=0.0.0.0
+      - "ES_JAVA_OPTS=-Xms512m -Xmx512m"
```
[the remark this answers is not in the document]
Comment (VasylKosiv) [reply docker-compose.yml]: The http module allows to expose Elasticsearch APIs over HTTP.
http.host     	Used to set the http.bind_host and the http.publish_host 
					http.bind_host The host address to bind the HTTP service to. 
					the http.publish_host The host address to publish for HTTP clients to connect to.
	In our case 0.0.0.0 means all IPv4 addresses on the local machine

The transport module is used for internal communication between nodes within the cluster.
transport.host  Used to set the transport.bind_host and the transport.publish_host
					transport.bind_host The host address to bind the transport service to.
					transport.publish_host The host address to publish for nodes in the cluster to connect to.
	In our case localhost			
					
network.host The node will bind to this hostname or IP address and publish (advertise) this host to other nodes in the cluster.
	In our case 0.0.0.0  will bind to all network interfaces.

ES_JAVA_OPTS Set the JVM heap size
				Xms min size of heap
				Xmx max size of heap
	In our case "ES_JAVA_OPTS=-Xms512m -Xmx512m" mean that size of heap will be 512 Mb

All this variables i took from https://docs.graylog.org/en/3.2/pages/installation/docker.html I dicided that veriables should be the same for good comunication between graylog and elasticsearch

**Pull request merged**